src/viewer-messaging.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  APP,
+  Messaging,
+  MessageType,
+  WindowPortEmulator,
+} from 'amp-viewer-messaging/messaging';
+import {log} from '../utils/log';
+
+
+const CHANNEL_OPEN_MSG = 'channelOpen';
+
+export class AmpViewerHost {
+
+  /**
+   * @param {!Window} win
+   * @param {!HTMLIFrameElement} ampIframe
+   * @param {string} frameOrigin
+   * @param {function(string, *, boolean):(!Promise<*>|undefined)} messageHandler
+   * @param {boolean=} opt_isWebview Should viewer initiate handshake w/ polling
+   * @param {boolean=} opt_isHandshakePoll
+   * looking at.
+   */
+  constructor(win, ampIframe, frameOrigin, messageHandler, opt_isWebview,
+    opt_isHandshakePoll) {
+    /** @const {!Window} */
+    this.win = win;
+    /** @private {!HTMLIFrameElement} */
+    this.ampIframe_ = ampIframe;
+    /** @private {function(string, *, boolean):(!Promise<*>|undefined)} */
+    this.messageHandler_ = messageHandler;
+    /** @const {boolean} */
+    this.isWebview_ = !!opt_isWebview;
+
+    if (this.isWebview_ || opt_isHandshakePoll) {
+      /** @private {number} */
+      this.pollingIntervalId_ = setInterval(this.initiateHandshake_.bind(
+        this, this.intervalCtr) , 1000); //poll every second
+    } else {
+      this.waitForHandshake_(frameOrigin);
+    }
+  }
+
+  /**
+   * @private
+   */
+  initiateHandshake_() {
+    log('initiateHandshake_');
+    if (this.ampIframe_) {
+      const channel = new MessageChannel();
+      let message = {
+        app: APP,
+        name: 'handshake-poll',
+      };
+      message = this.isWebview_ ? JSON.stringify(message) : message;
+      this.ampIframe_.contentWindow./*OK*/postMessage(
+        message, '*', [channel.port2]);
+
+      channel.port1.onmessage = function(e) {
+        const data = this.isWebview_ ? JSON.parse(e.data) : e.data;
+        if (this.isChannelOpen_(data)) {
+          this.win.clearInterval(this.pollingIntervalId_); //stop polling
+          log('messaging established!');
+          this.completeHandshake_(channel.port1, data.requestid);
+        } else {
+          this.messageHandler_(data.name, data.data, data.rsvp);
+        }
+      }.bind(this);
+    }
+  }
+
+  /**
+   * @param {string} targetOrigin
+   * @private
+   */
+  waitForHandshake_(targetOrigin) {
```

Comment:
Yes. I realize it's probably out of scope for this PR, but just it may be worth considering consolidating all messaging code there. Then the Viewer can instantiate a channel without worrying about the handshake details.